resources/todomvc/architecture-examples/react/src/todo/components/Input.jsx

```diff
@@ -0,0 +1,49 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+import { useSanitizer } from "../hooks/use-sanitizer";
+import { useValidators } from "../hooks/use-validators";
+
+export const Input = ({ onSubmit, placeholder, label, defaultValue, onBlur }) => {
+    const { sanitize } = useSanitizer();
+    const { hasValidMin } = useValidators();
+
+    const handleSubmit = (e) => {
+        // trim whitespaces
+        const value = e.target.elements["todo-input"].value.trim();
+        e.preventDefault();
+
+        // enforce 2 chars min
+        if (!hasValidMin(value, 2)) {
+            return;
+        }
+
+        // sanitize input and submit
+        onSubmit(sanitize(value));
+
+        e.target.reset();
+    };
+
+    const handleBlur = (e) => {
+        if (onBlur) {
+            onBlur();
+        }
+    };
+
+    return (
+        <form onSubmit={handleSubmit}>
+            <input className="new-todo" id="todo-input" type="text" data-testid="text-input" autoFocus placeholder={placeholder} defaultValue={defaultValue} onBlur={handleBlur} />
```

Comment:
In react I believe it's usually advised to implement as a "controlled input", where the value is kept in the local state. Here instead you implemented an uncontrolled input. I believe this works and I'll leave it up to you to decide what style you prefer, but I wanted to point it out. For example a controlled component might make it easier to access the value in `handleSubmit`.

Controlled components: https://reactjs.org/docs/forms.html#controlled-components
Uncontrolled components: https://reactjs.org/docs/uncontrolled-components.html

This is the old doc, it looks like that the new documentation website is less assertive about which style to use.

Reply:
yes - I used a form before 😄 
It does solve some problems, but even with a form I honestly wouldn't keep local state.
I could still read the value from the event (e.target.elements["todo-input"].value).